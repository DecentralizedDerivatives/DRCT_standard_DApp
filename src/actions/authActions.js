import { web3 } from '../ethereum';

import { GET_ERRORS, SET_USER_CONNECTED } from './types';

// Check connection status 
export const checkConnection = () => dispatch => {
  try {
    const accounts = await web3.eth.getAccounts();
    const network = await web3.eth.net.getId();
    
    dispatch({
      type: SET_USER_CONNECTED,
      payload: (accounts.length && network === 4)  
      })
    }
  } catch(err) {
    dispatch({
      type: GET_ERRORS,
      payload: err.response.data
    })
  }
}
