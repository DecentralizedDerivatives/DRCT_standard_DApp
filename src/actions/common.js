import { Oracle } from '../ethereum';

const moment = require('moment');

export const getStartDatePrice = async (oracleAddress, startDate) => {
  let dt = startDate ? moment(startDate, 'MM/DD/YYYY').utc().startOf('day') : null;
  let date = dt.format('x') / 1000
  const oracle = await Oracle.at(oracleAddress)
  var data = await oracle.retrieveData(date)
  if (!data || !data.c || !data.c[0]) { return 0 }
  let val = parseFloat('0.00' + data.c[0].toString())
  // console.log('val', val)
  let multiplier = '1e' + (data.e).toString()
  let adjustedValue = val * (Number(multiplier))
  // console.log('adjustedValue', adjustedValue)
  return adjustedValue
}
