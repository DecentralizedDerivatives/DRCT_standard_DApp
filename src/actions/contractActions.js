import { Factory, Exchange, DRCT, Oracle } from '../ethereum';
import {
  SET_CONTRACT_DETAILS,
  SET_CONTRACT_OPEN_DATES,
  SET_CONTRACT_START_PRICE,
  SET_ORDERBOOK,
  SET_FETCH_IN_PROGRESS,
  REMOVE_FETCH_IN_PROGRESS,
  SET_FETCHING_ERROR,
  SET_RECENT_TRADES
} from './types';

import FactoryProvider from '../factoryProvider';
const moment = require('moment');

export const getContractDetails = (symbol) => async dispatch => {
  try {
    const provider = FactoryProvider.getFromSymbol(symbol);
    // const factory = await Factory.at(provider && provider.address ? provider.address : '');
    // const response = await factory.getVariables();
    const details = {
      contractAddress: provider.address,
      contractDuration: provider.duration,
      contractMultiplier: provider.multiplier,
      oracleAddress: provider.oracle
      // contractAddress: response[0],
      // contractDuration: response[1].c[0],
      // contractMultiplier: response[2].c[0],
      // oracleAddress: response[3]
    };

    dispatch({
      type: SET_CONTRACT_DETAILS,
      payload: details
    });
    return details;
  } catch (err) {
    dispatch({
      type: SET_FETCHING_ERROR,
      payload: err.message.split('\n')[0]
    });
  }
};

export const getStartDatePrice = (oracleAddress, startDate) => async dispatch => {
  try {
    const oracle = await Oracle.at(oracleAddress)
    var data = await oracle.retrieveData(startDate);
    dispatch({
      type: SET_CONTRACT_START_PRICE,
      payload: data.c[0]
    });
  } catch (err) {
    dispatch({
      type: SET_FETCHING_ERROR,
      payload: err.message.split('\n')[0]
    });
  }
}
export const getOrderBook = (isSilent) => async dispatch => {
  try {
    if (!isSilent) { dispatch({ type: SET_FETCH_IN_PROGRESS, payload: SET_ORDERBOOK }); };
    var staticAddresses = FactoryProvider.getStaticAddresses();
    const exchange = await Exchange.at(staticAddresses.exchange);
    let numBooks = await exchange.getBookCount();
    var factories = FactoryProvider.factories();
    let _allrows = [];
    for (let i = 0; i < numBooks; i++) {
      let book = await exchange.openBooks(i);
      // console.log('book', book)
      for (var p = 0; p < factories.length; p++) {
        const factory = await Factory.at(factories[p].address);
        // console.log('factory', factories[p].symbol);
        let tokenDate = await factory.token_dates.call(book);
        // console.log('tokenDate', tokenDate)
        if (tokenDate.c[0] === 0) { continue }
        let orders = await exchange.getOrders(book);
        // console.log('orders', orders)
        for (let j = 0; j < orders.length; j++) {
          if (orders[j].c[0] > 0) {
            let order = await exchange.getOrder(orders[j].c[0]);
            // console.log('order', order)
            let tokenType = (await factory.getTokenType(order[3])).c[0];
            let date = new Date(tokenDate.c[0] * 1000);
            let startDate = new Date();
            startDate.setDate(date.getDate() + 1);
            var todayMinusSixDays = new Date();
            todayMinusSixDays.setDate(todayMinusSixDays.getDate() - 6);
            if (date > todayMinusSixDays) {
              let orderDate = startDate.getUTCMonth() + 1 + '/' +
                startDate.getUTCDate() + '/' + startDate.getUTCFullYear();
              var precisePrice = parseFloat(order[1].c[0]/10000).toFixed(5);
              _allrows.push({
                orderId: orders[j].c[0].toString(),
                address: order[3],
                price: precisePrice,
                quantity: order[2].c[0].toString(),
                date: orderDate.toString(),
                symbol: factories[p].symbol,
                tokenType: tokenType === 1 ? 'Short' : 'Long'
               });
            }
          }
        }
      }
      _allrows.sort(function (a, b) {
        return a.orderId - b.orderId;
      });
    }
    dispatch({
      type: SET_ORDERBOOK,
      payload: _allrows
    });
    dispatch({ type: REMOVE_FETCH_IN_PROGRESS, payload: SET_ORDERBOOK });
  } catch (err) {
    dispatch({
      type: SET_FETCHING_ERROR,
      payload: 'Order Book: ' + err.message.split('\n')[0]
    });
  }
};

export const getRecentTrades = (isSilent) => async dispatch => {
  try {
    if (!isSilent) { dispatch({ type: SET_FETCH_IN_PROGRESS, payload: SET_RECENT_TRADES }); };
    var staticAddresses = FactoryProvider.getStaticAddresses();
    const exchange = await Exchange.at(staticAddresses.exchange);

    let transferEvent = await exchange.Sale(
      {},
      { fromBlock: 0, toBlock: 'latest' }
    );

    transferEvent.get(async function (err, events) {
      // console.log('events', events);
      var trades = [];
      if (events.length > 0) {
        for (let i = events.length - 1; i >= Math.max(events.length - 10, 0); i--) {
          var token = events[i].args['_token'].toString();
          // console.log('token', token);
          var drct = DRCT.at(token);
          // console.log('drct', drct);
          var factoryAddress = await drct.getFactoryAddress();
          const factory = await Factory.at(factoryAddress);
          let tokenType = (await factory.getTokenType(token)).c[0];
          // console.log('factoryAddress', factoryAddress);
          var provider = FactoryProvider.getFromAddress(factoryAddress);
          var precisePrice = parseFloat(events[i].args['_price']/1e18).toFixed(5);
          trades.push({
            address: token,
            volume: events[i].args['_amount'].toString(),
            price: precisePrice,
            contractDuration: provider && provider.duration ? provider.duration : 0,
            contractMultiplier: provider && provider.multiplier ? provider.multiplier : 0,
            symbol: provider && provider.symbol ? provider.symbol : '??',
            tokenType: tokenType === 1 ? 'Short' : 'Long'
          });
        }
      }
      dispatch({
        type: SET_RECENT_TRADES,
        payload: trades
      });
      dispatch({ type: REMOVE_FETCH_IN_PROGRESS, payload: SET_RECENT_TRADES });
    });
  } catch (err) {
    dispatch({
      type: SET_FETCHING_ERROR,
      payload: 'Recent Trades: ' + err.message.split('\n')[0]
    });
  }
};

export const getContractOpenDates = (currencyAddress) => async dispatch => {
  try {
    const factory = await Factory.at(currencyAddress);
    // console.log(factory);
    let openDates = {};
    const numDates = await factory.getDateCount();

    for (let i = 0; i < numDates; i++) {
      const startDate = (await factory.startDates.call(i)).c[0];
      let formattedDate = new Date(startDate * 1000);
      var momentDate = moment(formattedDate);
      var pastDate = moment().subtract(5, 'days').startOf('day');
      if (momentDate.isBefore(pastDate)) { continue }
      formattedDate = formattedDate.getUTCMonth() + 1 + '/' +
        formattedDate.getUTCDate() + '/' +
        formattedDate.getUTCFullYear();
      openDates[startDate] = formattedDate;
    }
    dispatch({
      type: SET_CONTRACT_OPEN_DATES,
      payload: openDates
    });
  } catch (err) {
    dispatch({
      type: SET_FETCHING_ERROR,
      payload: 'Contract Open Dates: ' + err.message.split('\n')[0]
    });
  }
};
