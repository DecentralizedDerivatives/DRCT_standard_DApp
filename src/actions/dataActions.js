import api from '../api';
import { SET_PRICECHART_DATA, SET_FETCHING_ERROR } from './types';

export const getPriceChartData = type => async dispatch => {
  try {
    const data = await api[type].get();
      
    dispatch({
      type: SET_PRICECHART_DATA,
      payload: data 
    });      

  } catch (err) {
    dispatch({
      type: SET_FETCHING_ERROR,
      payload: 'PriceChart: ' + err.message.split('\n')[0]
    });
  }
};
