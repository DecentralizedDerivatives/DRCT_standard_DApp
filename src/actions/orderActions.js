import { Factory, Exchange, DRCT, UserContract } from '../ethereum';
import {
  SET_ORDER_DETAILS,
  SET_BUY_ORDER_RECEIPT,
  SET_BUY_ORDER_ERROR,
  SET_UNLIST_ORDER_RECEIPT,
  SET_UNLIST_ORDER_ERROR,
  SET_CONTRACT_CREATED,
  SET_CONTRACT_ERROR,
  SET_CONTRACT_FUNDED,
  SET_SEND_FUNDS_ERROR,
  SET_LIST_ORDER,
  SET_LIST_ORDER_ERROR,
  SET_LIST_ORDER_APPROVED,
  SET_LIST_ORDER_APPROVE_ERROR,
  SET_PROCESSING,
  SET_FETCHING_ERROR
} from './types';

import FactoryProvider from '../factoryProvider';

//TODO: This function makes no sense.  Why return an array of strings??
//Should be a single object with order details.  Right??
export const getOrderDetails = orderId => async dispatch => {
  try {
    if (!orderId) {
      return dispatch({type: SET_ORDER_DETAILS, payload: ''});
    }
    var factories = FactoryProvider.factories();
    var staticAddresses = FactoryProvider.getStaticAddresses();
    const exchange = await Exchange.at(staticAddresses.exchange);
    let order = await exchange.getOrder(orderId);
    let _allrows = [];
    for (var i = 0; i < factories.length; i++) {
      const factory = await Factory.at(factories[i].address);
      var orderDetail = await getOrderDetailForFactory(factory, order, orderId);
      _allrows.push(orderDetail);
    }
    dispatch({
      type: SET_ORDER_DETAILS,
      payload: _allrows
    });
  } catch (err) {
    dispatch({
      type: SET_FETCHING_ERROR,
      payload: 'Get Order Details: ' + err.message.split('\n')[0]
    });
  }
};

const getOrderDetailForFactory = async (factory, order, orderId) => {
  let display = [];
  let date = await factory.token_dates.call(order[3]);
  date = new Date(date * 1000);
  date =
    date.getMonth() + 1 + '/' + date.getDate() + '/' + date.getFullYear();
  display =
    orderId.toString() +
    '(' +
    order[3] +
    order[1].c[0].toString() +
    '/' +
    order[2].c[0].toString() +
    '/' +
    date.toString() +
    ')';
  return display;
};

export const sendBuyOrder = (orderId, account) => async dispatch => {
  dispatch(setProcessing(true));
  dispatch({type: SET_BUY_ORDER_ERROR, payload: null });
  try {
    orderId = parseInt(orderId, 10);
    var staticAddresses = FactoryProvider.getStaticAddresses();
    const exchange = await Exchange.at(staticAddresses.exchange);
    const order = await exchange.getOrder(orderId);
    const _value = order[1];
    const response = await exchange.buy(orderId, {
      from: account,
      gas: 400000,
      value: _value
    });
    // console.log('RESPONSE', response)
    dispatch({
      type: SET_BUY_ORDER_RECEIPT,
      payload: {
        id: response.tx,
        orderId: orderId
      }
    });
  } catch (err) {
    // console.log('err', err.message)
    dispatch({
      type: SET_BUY_ORDER_ERROR,
      payload: err.message.split('\n')[0]
    });
  }
  dispatch(setProcessing(false));
};

export const sendUnlistOrder = (orderId, account) => async dispatch => {
  dispatch(setProcessing(true));

  try {
    var staticAddresses = FactoryProvider.getStaticAddresses();
    const exchange = await Exchange.at(staticAddresses.exchange);
    const response = await exchange.unlist(orderId, {
      from: account,
      gas: 200000
    });

    dispatch({
      type: SET_UNLIST_ORDER_RECEIPT,
      payload: {
        id: response.tx,
        orderId: orderId
      }
    });
  } catch (err) {
    dispatch({
      type: SET_UNLIST_ORDER_ERROR,
      payload: err.message.split('\n')[0]
    });
  }

  dispatch(setProcessing(false));
};

export const sendListOrder = (formValues, account) => async dispatch => {
  dispatch(setProcessing(true));

  let { token, tokenAmount, price } = formValues;

  try {
    var staticAddresses = FactoryProvider.getStaticAddresses();
    const exchange = await Exchange.at(staticAddresses.exchange);

    const response = await exchange.list(token, tokenAmount, price * 1e18, {
      from: account,
      gas: 400000
    });

    dispatch({
      type: SET_LIST_ORDER,
      payload: {
        id: response.tx,
        price
      }
    });
  } catch (err) {
    dispatch({
      type: SET_LIST_ORDER_ERROR,
      payload: err.message.split('\n')[0]
    });
  }

  dispatch(setProcessing(false));
};

export const sendApproveOrder = (approveDetails, account) => async dispatch => {
  dispatch(setProcessing(true));

  let { token, tokenAmount } = approveDetails;

  try {
    var staticAddresses = FactoryProvider.getStaticAddresses();
    const exchange = await Exchange.at(staticAddresses.exchange);
    const drct = await DRCT.at(token);
    const allowance = await drct.allowance(account, exchange.address)
    const approvedAmount = allowance && allowance.c && allowance.c.length > 0 ? allowance.c[0] : 0
    let listedAmount = tokenAmount
    if (Number(tokenAmount) < Number(approvedAmount)) {
      const ordersList = await exchange.getUserOrders(account)
      for (let ord of ordersList) {
        let order = await exchange.getOrder(ord.c[0])
        if (order[3] === token) {
          listedAmount += order[2].c[0]
        }
      }
    }
    if (Number(listedAmount) > Number(approvedAmount)) {
      await drct.approve(exchange.address, tokenAmount, {
        from: account,
        gas: 100000
      });
    }
    dispatch({
      type: SET_LIST_ORDER_APPROVED,
      payload: {
        token,
        tokenAmount,
        approved: true
        // approveTx: response.tx
      }
    });
  } catch (err) {
    console.log('APPROVE ERROR', err)
    dispatch({
      type: SET_LIST_ORDER_APPROVE_ERROR,
      payload: err.message.split('\n')[0]
    });
  }

  dispatch(setProcessing(false));
};

export const sendCreateContractOrder = (
  formValues,
  account
) => async dispatch => {
  dispatch(setProcessing(true));
  
  const { duration, currency, startDate } = formValues;
  var amount = formValues.amount / 2;
  try {
    const factory = await Factory.at(currency);

    let userAddress = await factory.user_contract.call();
    const userContract = await UserContract.at(userAddress);

    let fundingAmount = 1e18 * amount;
    console.log('Variables',startDate,fundingAmount)
    const response = await userContract.Initiate(startDate, fundingAmount, {
      from: account,
      gas: 1100000,
      value: fundingAmount * 2
    });

    dispatch({
      type: SET_CONTRACT_FUNDED,
      payload: response.tx
    });
  } catch (err) {
    dispatch({
      type: SET_SEND_FUNDS_ERROR,
      payload: err.message.split('\n')[0]
    });
  }
  dispatch(setProcessing(false));
};

export const setProcessing = status => {
  return {
    type: SET_PROCESSING,
    payload: status
  };
};
