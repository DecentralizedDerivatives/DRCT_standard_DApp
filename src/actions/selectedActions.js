import { SET_SELECTED_TOKEN, CLEAR_SELECTED_TOKEN } from './types';

export const setSelectedToken = token => dispatch => {
  dispatch({
    type: SET_SELECTED_TOKEN,
    payload: token
  });
};

export const clearSelectedToken = token => dispatch => {
  dispatch({
    type: CLEAR_SELECTED_TOKEN,
    payload: ''
  });
};
