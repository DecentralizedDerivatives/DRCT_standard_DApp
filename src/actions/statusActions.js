import { web3 } from '../ethereum';

import {
  SET_FETCHING_ERROR,
  SET_CONNECTION_STATUS,
  SHOW_CONNECTION_MODAL,
  SET_PROCESSING
} from './types';

// Check connection status
export const checkUserConnection = () => async dispatch => {
  try {
    const accounts = await web3.eth.getAccounts();
    const network = await web3.eth.net.getId();
    console.log("NETWORK",network);
    dispatch({
      type: SET_CONNECTION_STATUS,
      payload: {
        metamask: Boolean(accounts.length),
        network: network
      }
    });
  } catch (err) {
    dispatch({
      type: SET_FETCHING_ERROR,
      payload: err.message.split('\n')[0]
    });
  }
};

// Set connection modal status
export const showConnectionModal = status => {
  return {
    type: SHOW_CONNECTION_MODAL,
    payload: status
  };
};

export const setProcessing = status => {
  return {
    type: SET_PROCESSING,
    payload: status
  };
};
