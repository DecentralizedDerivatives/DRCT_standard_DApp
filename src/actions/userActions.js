import { Factory, Exchange, web3, DRCT, Wrapped } from '../ethereum';
import {
  SET_USER_ACCOUNT,
  SET_USER_BALANCE,
  SET_USER_TRANSACTIONS,
  SET_USER_POSITIONS,
  SET_USER_TOKENS,
  SET_SELECTED_TOKEN,
  SET_USER_ORDERS,
  // SET_USER_ORDER_LABELS,
  SET_CURRENT,
  SET_PROCESSING,
  SET_FETCHING_ERROR,
  SET_CASHOUT_RECEIPT,
  SET_CASHOUT_ERROR
} from './types';

import FactoryProvider from '../factoryProvider';

export const getUserAccount = () => async dispatch => {
  dispatch(setProcessing(true));
  try {
    const userAccounts = await web3.eth.getAccounts();

    dispatch({
      type: SET_USER_ACCOUNT,
      payload: userAccounts[0]
    });
  } catch (err) {
    dispatch({
      type: SET_FETCHING_ERROR,
      payload: err.message.split('\n')[0]
    });
  }

  dispatch(setProcessing(false));
};

export const getUserBalance = () => async dispatch => {
  dispatch(setProcessing(true));
  try {
    const wrapped = await Wrapped.deployed();
    const accounts = await web3.eth.getAccounts();
    let _res = await wrapped.balanceOf(accounts[0]);

    dispatch({
      type: SET_USER_BALANCE,
      payload: _res.c[0]
    });
  } catch (err) {
    dispatch({
      type: SET_FETCHING_ERROR,
      payload: err.message.split('\n')[0]
    });
  }

  dispatch(setProcessing(false));
};

export const getUserTransactions = userAccount => async dispatch => {
  dispatch(setProcessing(true));
  try {
    //const exchange = await Exchange.deployed();
    var factories = FactoryProvider.factories();
    var transactions = [];
    for (var i = 0; i < factories.length; i++) {
      const factory = await Factory.at(factories[i].address);
      var events = await getContractCreationEvents(factory, userAccount);
      transactions = transactions.concat(events);
    };
    //     //Sale: await exchange.Sale({}, {fromBlock:0, toBlock: 'latest'}),
    //     //OrderPlaced: await exchange.OrderPlaced({}, {fromBlock:0, toBlock: 'latest'}),
    //     //OrderRemoved: await exchange.OrderRemoved({}, {fromBlock:0, toBlock: 'latest'}),
    //     ContractCreation: await factory.ContractCreation(
    //       {},
    //       { fromBlock: 0, toBlock: 'latest' }
    //     )
    //     //Transfer: await drct.Transfer({}, {fromBlock:0, toBlock: 'latest'}),
    //     //Approval: await drct.Approval({}, {fromBlock:0, toBlock: 'latest'})
    dispatch({
      type: SET_USER_TRANSACTIONS,
      payload: transactions
    });
  } catch (err) {
    dispatch({
      type: SET_FETCHING_ERROR,
      payload: 'User Transactions: ' + err.message.split('\n')[0]
    });
  }

  dispatch(setProcessing(false));
};

const getContractCreationEvents = async (factory, userAccount) => {
  let transferEvent = await factory.ContractCreation(
    {},
    { fromBlock: 0, toBlock: 'latest' }
  );
  return new Promise((resolve, reject) => {
    transferEvent.get(function (err, logs) { // .get() does not support async/await
      try {
        let trades = [];
        for (let i = logs.length - 1; i >= Math.max(logs.length - 10, 0); i--) {
          if (logs[i].args['_sender'].toUpperCase() === userAccount.toUpperCase()) {
            trades.push({title: 'ContractCreation', hash: logs[i].transactionHash});
          }
        }
        resolve(trades);
      } catch (err) {
        reject(err);
      }
    });
  });
}

export const getUserPositions = userAccount => async dispatch => {
  dispatch(setProcessing(true));
  try {
    var factories = FactoryProvider.factories();
    var positions = []
    for (var i = 0; i < factories.length; i++) {
      var data = await getPositionsForFactory(factories[i], userAccount);
      positions = positions.concat(data);
    };
    dispatch({
      type: SET_USER_POSITIONS,
      payload: positions
    });
  } catch (err) {
    dispatch({
      type: SET_FETCHING_ERROR,
      payload: 'User Positions: ' + err.message.split('\n')[0]
    });
  }

  dispatch(setProcessing(false));
};

const getPositionsForFactory = async (provider, userAccount) => {
  const factory = await Factory.at(provider.address);
  let positions = [];
  const numDates = await factory.getDateCount();
  for (let i = 0; i < numDates; i++) {
    const startDate = (await factory.startDates.call(i)).c[0];
    const tokenAddresses = await factory.getTokens(startDate);
    for (let p = 0; p < tokenAddresses.length; p++) {
      let drct = await DRCT.at(tokenAddresses[p]);
      let balance = await drct.balanceOf(userAccount);
      if (balance.c[0] > 0) {
        let date = new Date(startDate * 1000);
        date = date.getUTCMonth() + 1 + '/' +
          date.getUTCDate() + '/' +
          date.getUTCFullYear();
        positions.push({
          address: tokenAddresses[p],
          balance: balance.c[0].toString(),
          date: date.toString(),
          symbol: provider.symbol,
          contractDuration: 0, // TODO: this.state.contractDuration,
          contractMultiplier: 0, // TODO: this.state.contractMultiplier
        });
      }
    }
  }
  return positions;
}

export const getUserTokenPositions = userAccount => async dispatch => {
  dispatch(setProcessing(true));
  try {
    var factories = FactoryProvider.factories();
    var tokens = []
    for (var i = 0; i < factories.length; i++) {
      const factory = await Factory.at(factories[i].address);
      var data = await getTokenPositionsForFactory(factory, userAccount);
      tokens = tokens.concat(data);
    }
    dispatch({
      type: SET_USER_TOKENS,
      payload: tokens
    });
    dispatch({
      type: SET_SELECTED_TOKEN,
      payload: {
        selectedToken: tokens.length > 0 ? tokens[0] : null
      }
    });
  } catch (err) {
    dispatch({
      type: SET_FETCHING_ERROR,
      payload: 'User Token Positions: ' + err.message.split('\n')[0]
    });
  }

  dispatch(setProcessing(false));
};

const getTokenPositionsForFactory = async (factory, userAccount) => {
  const numDates = await factory.getDateCount();
  let tokens = [];
  for (let i = 0; i < numDates; i++) {
    const startDates = (await factory.startDates.call(i)).c[0];
    const tokenAddresses = await factory.getTokens(startDates);
    let _date = new Date(startDates * 1000);
    _date = _date.getMonth() + 1 + '/' +
      _date.getDate() + '/' +
      _date.getFullYear();

    for (let p = 0; p < tokenAddresses.length; p++) {
      const drct = await DRCT.at(tokenAddresses[p]); //Getting contract
      const balance = (await drct.balanceOf(userAccount)).c[0]; //Getting balance of token
      if (balance > 0) {
        tokens.push(`${tokenAddresses[p]}(${balance}/${_date})`); //Pushing token address + balance/date
      };
    }
  }
  return tokens;
};
export const getUserOrders = userAccount => async dispatch => {
  dispatch(setProcessing(true));
  try {
    var factories = FactoryProvider.factories();
    var orders = []
    for (var i = 0; i < factories.length; i++) {
      var data = await getOrdersForFactory(factories[i], userAccount);
      orders = orders.concat(data);
    }
    dispatch({
      type: SET_USER_ORDERS,
      payload: {
        userOrders: orders
      }
    });
    dispatch({
      type: SET_CURRENT,
      payload: {
        selectedToken: orders.length > 0 ? orders[0].row : null,
        selectedOrderID: orders.length > 0 ? orders[0].id : null
      }
    });
  } catch (err) {
    dispatch({
      type: SET_FETCHING_ERROR,
      payload: 'User Orders: ' + err.message.split('\n')[0]
    });
  }

  dispatch(setProcessing(false));
};
const getOrdersForFactory = async (factory, userAccount) => {
  const exchange = await Exchange.deployed();
  const books = await exchange.getUserOrders.call(userAccount); //Gets all listed order ids
  const allOrders = []; //Contains all information for each order
  for (let i = 0; i < books.length; i++) {
    const order = {};
    order.id = books[i].c[0];
    order.info = await exchange.getOrder(order.id); //Getting order info by order Id (returns array);
    order.owner = order.info[0];
    order.price = order.info[1].c[0] / 10000; //divided by 10000 to fix offset
    order.owned = order.info[2].c[0];
    order.address = order.info[3];
    var date = await factory.token_dates.call(order.address);
    date = new Date(date * 1000);
    order.date = date.getMonth() + 1 + '/' +
      date.getDate() + '/' +
      date.getFullYear();
    order.row = `${order.address}(${order.owned}/${order.date})`;
    allOrders.push(order);
  }
  return allOrders;
}

export const sendCashOutRequest = (amount, account) => async dispatch => {
  dispatch(setProcessing(true));

  try {
    const wrapped = await Wrapped.deployed();
    const response = await wrapped.withdraw(amount, {
      from: account,
      gas: 4000000
    });

    dispatch({
      type: SET_CASHOUT_RECEIPT,
      payload: {
        id: response.tx,
        amount: amount
      }
    });
  } catch (err) {
    dispatch({
      type: SET_CASHOUT_ERROR,
      payload: err.message.split('\n')[0]
    });
  }

  dispatch(setProcessing(false));
};

export const setProcessing = status => {
  return {
    type: SET_PROCESSING,
    payload: status
  };
};
