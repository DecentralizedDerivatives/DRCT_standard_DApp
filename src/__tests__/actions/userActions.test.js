import * as userActions from '../../actions/userActions';

jest.mock('../../ethereum');
jest.mock('../../factoryProvider');
jest.mock('../../actions/common', () => ({
  getStartDatePrice: jest.fn(() => 1),
}));
jest.mock('../../api', () => ({
  eth: {
    get: jest.fn(() => [0, 1]),
  },
}));

import {
  web3,
  WrappedInstance,
  ExchangeInstance,
  FactoryInstance,
  DRCTInstance,
} from '../../ethereum';
import { getStartDatePrice } from '../../actions/common';

web3.eth.getAccounts.mockImplementation(() => ['0x000...']);

const store = mockStore({});

describe('userActions', () => {
  afterEach(() => {
    store.clearActions();
  });

  it('getUserAccount', async () => {
    await store.dispatch(userActions.getUserAccount());
    expect(store.getActions()).toMatchSnapshot();
  });

  it('getUserAccount error', async () => {
    web3.eth.getAccounts.mockImplementationOnce(() => undefined);
    await store.dispatch(userActions.getUserAccount());
    expect(store.getActions()).toMatchSnapshot();
  });

  it('getUserBalance', async () => {
    await store.dispatch(userActions.getUserBalance());
    expect(store.getActions()).toMatchSnapshot();
  });

  it('getUserBalance error', async () => {
    WrappedInstance.balanceOf.mockImplementationOnce(() => undefined);
    await store.dispatch(userActions.getUserBalance());
    expect(store.getActions()).toMatchSnapshot();
  });

  it('getUserTransactions', async () => {
    await store.dispatch(userActions.getUserTransactions('0x000...'));
    expect(store.getActions()).toMatchSnapshot();
  });

  it('getUserTransactions filters by account', async () => {
    await store.dispatch(userActions.getUserTransactions('0x111...'));
    expect(store.getActions()).toMatchSnapshot();
  });

  it('getUserTransactions error', async () => {
    FactoryInstance.ContractCreation.mockImplementationOnce(() => ({
      get: jest.fn(callback => callback('error')),
    }));
    await store.dispatch(userActions.getUserTransactions('0x000...'));
    expect(store.getActions()).toMatchSnapshot();
  });

  it('getUserTransactions silent', async () => {
    await store.dispatch(userActions.getUserTransactions('0x000...', true));
    expect(store.getActions()).toMatchSnapshot();
  });

  it('getUserPositions', async () => {
    await store.dispatch(userActions.getUserPositions());
    expect(store.getActions()).toMatchSnapshot();
  });

  it('getUserPositions error', async () => {
    FactoryInstance.getTokens.mockImplementationOnce(() => undefined);
    await store.dispatch(userActions.getUserPositions());
    expect(store.getActions()).toMatchSnapshot();
  });

  it('getUserPositions silent', async () => {
    await store.dispatch(userActions.getUserPositions('0x000...', true));
    expect(store.getActions()).toMatchSnapshot();
  });

  it('getUserPositions short token', async () => {
    FactoryInstance.getTokenType.mockImplementationOnce(() => ({ c: [1] }));
    await store.dispatch(userActions.getUserPositions());
    expect(store.getActions()).toMatchSnapshot();
  });

  it('getUserPositions no balance', async () => {
    DRCTInstance.balanceOf.mockImplementationOnce(() => ({ c: [0] }));
    await store.dispatch(userActions.getUserPositions());
    expect(store.getActions()).toMatchSnapshot();
  });

  it('getUserPositions no start price', async () => {
    getStartDatePrice.mockImplementationOnce(() => undefined);
    await store.dispatch(userActions.getUserPositions());
    expect(store.getActions()).toMatchSnapshot();
  });

  it('getUserTokenPositions', async () => {
    await store.dispatch(userActions.getUserTokenPositions());
    expect(store.getActions()).toMatchSnapshot();
  });

  it('getUserTokenPositions error', async () => {
    FactoryInstance.getTokens.mockImplementationOnce(() => undefined);
    await store.dispatch(userActions.getUserTokenPositions());
    expect(store.getActions()).toMatchSnapshot();
  });

  it('getUserOrders', async () => {
    await store.dispatch(userActions.getUserOrders());
    expect(store.getActions()).toMatchSnapshot();
  });

  it('getUserOrders error', async () => {
    ExchangeInstance.getUserOrders.mockImplementationOnce(() => undefined);
    await store.dispatch(userActions.getUserOrders());
    expect(store.getActions()).toMatchSnapshot();
  });

  it('sendCashOutRequest', async () => {
    await store.dispatch(userActions.sendCashOutRequest());
    expect(store.getActions()).toMatchSnapshot();
  });

  it('sendCashOutRequest error', async () => {
    WrappedInstance.withdraw.mockImplementationOnce(() => undefined);
    await store.dispatch(userActions.sendCashOutRequest());
    expect(store.getActions()).toMatchSnapshot();
  });

  it('setProcessing', async () => {
    await store.dispatch(userActions.setProcessing());
    expect(store.getActions()).toMatchSnapshot();
  });
});
