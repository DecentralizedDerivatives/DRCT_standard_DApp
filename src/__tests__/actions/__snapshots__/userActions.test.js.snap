// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`userActions getUserAccount 1`] = `
Array [
  Object {
    "payload": "Cannot read property '0' of undefined",
    "type": "SET_FETCHING_ERROR",
  },
]
`;

exports[`userActions getUserBalance 1`] = `
Array [
  Object {
    "payload": "Cannot read property '0' of undefined",
    "type": "SET_FETCHING_ERROR",
  },
]
`;

exports[`userActions getUserOrders 1`] = `
Array [
  Object {
    "payload": Array [
      Object {
        "address": "0x000...",
        "date": "NaN/NaN/NaN",
        "id": 0,
        "info": Array [
          Object {
            "c": Array [
              10,
            ],
          },
          Object {
            "c": Array [
              20,
            ],
          },
          Object {
            "c": Array [
              30,
            ],
          },
          "0x000...",
        ],
        "owned": 30,
        "owner": Object {
          "c": Array [
            10,
          ],
        },
        "price": 0.002,
        "row": "0x000...(30/NaN/NaN/NaN)",
      },
    ],
    "type": "SET_USER_ORDERS",
  },
]
`;

exports[`userActions getUserPositions 1`] = `
Array [
  Object {
    "payload": "SET_USER_POSITIONS",
    "type": "SET_FETCH_IN_PROGRESS",
  },
  Object {
    "payload": "User Positions: factory.getDateCount is not a function",
    "type": "SET_FETCHING_ERROR",
  },
]
`;

exports[`userActions getUserTokenPositions 1`] = `
Array [
  Object {
    "payload": "User Token Positions: factory.getDateCount is not a function",
    "type": "SET_FETCHING_ERROR",
  },
]
`;

exports[`userActions getUserTransactions 1`] = `
Array [
  Object {
    "payload": "SET_USER_TRANSACTIONS",
    "type": "SET_FETCH_IN_PROGRESS",
  },
  Object {
    "payload": "User Transactions: factory.ContractCreation is not a function",
    "type": "SET_FETCHING_ERROR",
  },
]
`;

exports[`userActions sendCashOutRequest 1`] = `
Array [
  Object {
    "payload": true,
    "type": "SET_PROCESSING",
  },
  Object {
    "payload": "Cannot read property '0' of undefined",
    "type": "SET_CASHOUT_ERROR",
  },
  Object {
    "payload": false,
    "type": "SET_PROCESSING",
  },
]
`;

exports[`userActions setProcessing 1`] = `
Array [
  Object {
    "payload": undefined,
    "type": "SET_PROCESSING",
  },
]
`;
